README
NAME
    IO::EventMux - Multiplexer for sockets, pipes and any other types of
    filehandles that you can set O_NONBLOCK on and does buffering for the
    user.

SYNOPSIS
      use IO::EventMux;

      my $mux = IO::EventMux->new();

      $mux->add($my_fh);

      while (1) {
        my $event = $mux->mux();

        # ... do something with $event->{type} and $event->{fh}
      }

DESCRIPTION
    This module provides multiplexing for any set of sockets, pipes, or
    whatever you can set O_NONBLOCK on. It can be useful for both server and
    client processes, but it works best when the application's main loop is
    centered around its "mux()" method.

    The file handles it can work with are either perl's own typeglobs or
    IO::Handle objects (preferred).

METHODS
  new()
    Constructs an IO::EventMux object.

  mux([$timeout])
    This method will block until ether an event occurs on one of the file
    handles or the $timeout (floating point seconds) expires. If the
    $timeout argument is not present, it waits forever. If $timeout is 0, it
    returns immediately.

    The return value is always a hash, which always has the key 'type',
    indicating what kind it is. It will also usually carry the 'fh' key,
    indicating what file handle the event happened on.

    The 'type' key can have the following values:

    timeout
        Nothing happened and timeout occurred.

    error
        An error occurred in connection with the file handle, such as
        "connection refused", etc.

    accepted
        A new client connected to a listening socket and the connection was
        accepted by EventMux. The listening socket file handle is in the
        'parent_fh' key. If the file handle is a unix domain socket the
        credentials of the user connection will be available in the keys;
        'pid', 'uid' and 'gid'.

    ready
        A file handle is ready to be written to, this can be use full when
        working with nonblocking connects so you know when the remote
        connection accepted the connection.

    accepting
        A new client is trying to connect to a listening socket, but the
        user code must call accept manually. This only happens when the
        ManualAccept option is set.

    read
        A socket has incoming data. If the socket's Buffered option is set,
        this will be what the buffering rule define.

        The data is contained in the 'data' key of the event hash. If recv()
        returned a sender address, it is contained in the 'sender' key and
        must be manually unpacked according to the socket domain, e.g. with
        "Socket::unpack_sockaddr_in()".

    read_last
        A socket last data before it was closed did not match the buffering
        rules, as defined by the IO::Buffered type given. he read_last type
        contains the result of a call to "read_last()" on the chosen buffer
        type.

        The default is not to return read_last and if no buffer is set read
        will contain this information.

    sent
        A socket has sent all the data in it's queue with the send call.
        This however does not indicate that the data has reached the other
        end, normally only that the data has reached the local buffer of the
        kernel.

    closing
        A file handle was detected to be have been closed by the other end
        or the file handle was set to be closed by the user. So EventMux
        stooped listening for events on this file handle. Event data like
        'Meta' is still accessible.

        The 'missing' key indicates the amount of data or packets left in
        the user space buffer when the file handle was closed. This does not
        indicate the amount of data received by the other end, only that the
        user space buffer left.

    closed
        A socket/pipe was disconnected/closed, the file descriptor, all
        internal references, and data store with the file handle was
        removed.

    can_write
        The ManualWrite option is set for the file handle, and "select()"
        has indicated that the handle can be written to.

    can_read
        The ManualRead option is set for the file handle, and "select()" has
        indicated that the handle can be read from.

  add($handle, [ %options ])
    Add a socket to the internal list of handles being watched.

    The optional parameters for the handle will be taken from the
    IO::EventMux object if not given here:

   Listen
    Defines if the file handle should be treated as a listening socket, the
    default is to auto detect this. I should not be necessary to set this
    value.

    The socket must be set up for listening, which is easily done with
    IO::Socket::INET:

      my $listener = IO::Socket::INET->new(
        Listen    => 5,
        LocalPort => 7007,
        ReuseAddr => 1,
      );

      $mux->add($listener);

   Type
    Either "stream" or "dgram". Should be auto detected in most cases.

    Defaults to "stream".

   ManualAccept
    If a connection comes in on a listening socket, it will by default be
    accepted automatically, and "mux()" will return a 'connect' event. If
    ManualAccept is set an 'accepting' event will be returned instead, and
    the user code must handle it itself.

      $mux->add($my_fh, ManualAccept => 1);

   ManualWrite
    By default EventMux handles nonblocking writing and you should use
    "$mux-"send($fh, $data)> or "$mux-"sendto($fh, $addr, $data)> to send
    your data, but if for some reason you send data yourself you can tell
    EventMux not to do writing for you and generate a 'can_write' event
    instead.

      $mux->add($my_fh, ManualWrite => 1);

    In both cases you can use "send()" to write data to the file handle.

    Note: If both ManualRead and ManualWrite is set, EventMux will not set
    the socket to nonblocking.

   ManualRead
    By default EventMux will handle nonblocking reading and generate a read
    event with the data, but if some reason you would like to do the reading
    yourself you can have EventMux generate a 'can_read' event for you
    instead.

      $mux->add($my_fh, ManualRead => 1);

    Never read or recv on the file handle. When the socket becomes readable,
    a "can_read()" event is returned.

    Note: If both ManualRead and ManualWrite is set, EventMux will not set
    the socket to nonblocking.

   ReadSize
    By default IO::EventMux will try to read 65536 bytes from the file
    handle, setting this options to something smaller might help make it
    easier for EventMux to be fair about how it returns it's event, but will
    also give more overhead as more system calls will be required to empty a
    file handle.

   Errors
    By default IO::EventMux will not deal with socket errors on non
    connected sockets such as a UDP socket in listening mode or where no
    peer has been defined. Or in other words whenever you use "sendto()" on
    socket. When enabling error handling, IO::EventMux sets the socket to
    collect errors with the MSG_ERRQUEUE option and collect errors with
    "recvmsg()" call.

    Errors are sent as error events with a little more information than
    normal, eg:

      $event = {
        data     => 'packet data',
        dst_port => 'destination port',
        from     => 'ip where the error is from',
        dst_ip   => 'destination ip',
      }

   Meta
    An optional scalar piece of metadata for the file handle. Can be
    retrieved and manipulated later with meta()

   Buffered
    IO::EventMux supports buffering of data before generating events, this
    can be used to only return events when a "complete" event is done. For
    this IO::EventMux uses IO::Buffered.

      # Would only return when a complete line 
      $mux->add($goodfh, Buffered => new IO::Buffered(Split => qr/\n/));

    Read more here: IO::Buffered

  listen()
    Wrapper around connect() with option (Listen => SOMAXCONN) set

  connect()
    Connect and add a socket to IO::EventMux, by using either URL syntax or
    IO::Socket Syntax. All options related to IO::EventMux is passed when
    calling add() on the new socket. Connect returns the new socket on
    completion.

    URL Syntax supports this format:

     * (tcp|udp)://HOST:PORT, Returns a udp of tcp socket.
     * (unix|unix_dgram)://path/file.sock, Returns a unix domain socket connection.

    For more information on how to use IO::Socket syntax look in
    IO::Socket::INET and IO::Socket::UNIX.

    Example of URL syntax; making a connection to localhost port 22

      my $fh = $mux->connect("tcp://127.0.0.1:22");

    Example of the same thing in IO::Socket Syntax;

      my $fh = $mux->connect(
        Proto => 'tcp',
        PeerAddr => '127.0.0.1',
        PeerPort => 22,
      );

  set()
    Set new options on a fh in IO::EventMux, currently only Buffered options
    is handled

  handles()
    Returns a list of file handles managed by this object.

  has_events()
    Returns true if there are pending events, or false otherwise

  type()
    Returns the socket type for a file handle

  class()
    Returns the socket class for a file handle

  meta($fh, [$newval])
    Set or get a piece of metadata on the filehandle. This can be any scalar
    value.

  remove($fh)
    Make EventMux forget about a file handle. The caller will then take over
    the responsibility of closing it.

  close($fh)
    Close a file handle. IO::EventMux will stop listing to both reads and
    writes on the file handle and return a "closing" event and on next "mux"
    call kill will be called, returning "closed" for the file handle.

    Note: All 'Meta' data associated with the file handle will be kept until
    the final 'closed' event is returned.

  kill($fh)
    Closes a file handle without giving time to finish any outstanding
    operations. Returns a 'closed' event, deletes all buffers and does not
    keep 'Meta' data.

    Note: Does not return the 'read_last' event.

  buflen($fh)
    Queries the length of the output buffer for this file handle. This only
    applies if ManualWrite is turned off, which is the default. For
    Type="dgram" sockets, it returns the number of datagrams in the queue.

    An application can use this method to see whether it should send more
    data or wait until the buffer queue is a bit shorter.

  recvdata($fh, $length)
    TODO: Queues @data to be written to the file handle $fh. Can only be
    used when ManualWrite is off (default).

  send($fh, @data)
    Queues @data to be written to the file handle $fh. Can only be used when
    ManualWrite is off (default).

    If the socket is of Type="stream"
        Returns true on success, undef on error. The data is sent when the
        socket becomes unblocked and a 'sent' event is posted when all data
        is sent and the buffer is empty. Therefore the socket should not be
        closed until "buflen($fh)" returns 0 or a sent request has been
        posted.

    If the socket is of Type="dgram"
        Each item in @data will be sent as a separate packet. Returns true
        on success and undef on error.

  sendto($fh, $to, @data)
    Like "send()", but with the recepient $to as a packed sockaddr
    structure, such as the one returned by "Socket::pack_sockaddr_in()".
    Only for Type="dgram" sockets.

      $mux->sendto($my_fh, pack_sockaddr_in($port, inet_aton($ip)), $data);

  push_event($event)
    Push event on queue

  nonblock($fh)
    Puts socket into nonblocking mode.

  socket_creds($fh)
    Return credentials on UNIX domain sockets.

  socket_type($fh)
    Return socket type.

  socket_listening($fh)
    Check if the socket is set to listening mode

  recroak()
    Helper function to rethrow croaks

  socket_errors
    Dummy sub that casts an error if the IO::EventMux::Socket::MsgHdr is not
    installed and the Errors option is used

  NOTES
    Working with PIPE's: When the other end of a pipe closes it's end,
    signals can get thrown. To handle this a signal handler needs to be
    defined:

      # Needed when writing to a broken pipe 
      $SIG{PIPE} = sub {
         croak "Broken pipe";
     };

    Getting rid of 'Filehandle ... opened only for output'

      # Needed as sysread() throws warnings when STDIN gets closed by the child
      $SIG{__WARN__} = sub {
         croak @_;    
      };

AUTHOR
    Jonas Jensen <jonas@infopro.dk>, Troels Liebe Bentsen
    <troels@infopro.dk>

COPYRIGHT AND LICENCE
    Copyright 2006-2008: Troels Liebe Bentsen Copyright 2006-2007: Jonas
    Jensen

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



ChangeLog
Version 2.02 ()
 * mux() will now return undef when the event queue is empty, there are no 
   handles to watch and there is no timeouts to wait for, this was done to make
   it easier to drop out of the loop when there are no chance that any events 
   will be generated.
 * Formating fixes from Frank Wiegand

Version 2.01 (Sun Apr 15 2009)
 * Remove IO::Epoll as default mistake

Version 2.01 (Sun Apr 15 2009)
 * Added pluggable EventLoop mechanism and added support for IO::Epool 
 * Fixed tests so there is no more hard dependence on IO::Buffered
 * Small fix for IO::EventMux::Dump eating options.

Version 2.00 (Sun Mar 22 2009)
 * Added IO::EventMux::Dump from José Micó
 * Added new function has_events as suggested by José Micó
 * Added new connect() call as shorthand for creating socket/adding it
 * Use Errno instead of POSIX as this takes less mem and is faster to load.
 * Added support for auto detecting if a socket is in listening mode or not.
   Listen => 1 on add() is no longer needed.
 * Rename "error_num" to "errno" in error events
 * Add new "Errors" options to add() that returns ICMP errors as error events.
 * Close no longer waits for send buffer to empty
 * Drop support for timeouts pr. file handle
 * Convert to new IO::Buffered and drop support for old Buffered API.
 * Added a lot more error handling, IO::EventMux now croaks when given wrong
   in a lot of cases.
 * Add credential information to unix sockets on accepted event.
 * Mux now croaks on negativ timeout value

Version 1.02 (Sun Dec 30 2007)
 * Make push_event public, so fake events can be pushed to the queue.
 * Fixed bug in fixed size buffering
 * Fix all Perl::Critic warning
 * Add experimental HTTP buffer type
 * Fixed bug in socket type detection
 * Croak when sending to a $fh that is not ours

Version 1.01 (Tue Sep 4 2007)
 * Fixed test that was time dependent


IO/EventMuxOld.pm
package IO::EventMuxOld;

=head1 NAME

IO::EventMux - Multiplexer for sockets or pipes with line buffering

=head1 SYNOPSIS

  use IO::EventMux;

  my $mux = IO::EventMux->new(
    Listen => $my_fd
  );

  while (1) {
    my $event = $mux->mux();

    # ... do something with $event->{type} and $event->{fh}
  }

=head1 DESCRIPTION

This module provides multiplexing for any set of sockets, pipes, or whatever
you can set O_NONBLOCK on.  It can be useful for both server and client
processes, but it works best when the application's main loop is centered
around its C<mux()> method.

The file handles it can work with are either perl's own typeglobs or IO::Handle
objects (preferred).

=head1 METHODS

=cut

use strict;
use warnings;

use IO::Select;
use IO::Socket;
use Socket;
use Carp qw(carp cluck);
use Fcntl;
use POSIX qw(EWOULDBLOCK ENOTCONN EAGAIN F_GETFL F_SETFL O_NONBLOCK);

=head2 B<new(%options = ())>

Constructs an IO::EventMux object, setting a set of options:

=over

=item Listen

Can be either a socket or an array of sockets.  These sockets must be set up
for listening, which is easily done with IO::Socket::INET sockets:

  my $listener = IO::Socket::INET->new(
    Listen    => 5,
    LocalPort => 7007,
    ReuseAddr => 1,
  );

=item ManualAccept

If a connection comes in on a listening socket, it will by default be accepted
automatically, and mux() will return a 'connect' event.  If ManualAccept is
set, an 'accept' event will be returned instead, and the user code must handle
it itself.

=item LineBuffered

Global default for new file handles.  See C<add()> for a description. Default
is off.

=item ManualWrite

Global default for new file handles.  See C<add()> for a description. Default
is off.

=item ManualRead

Global default for new file handles.  See C<add()> for a description. Default
is off.

=item Type

Global default for new file handles.  See C<add()> for a description. Default
is off.

=back

=cut

sub new {
    my ($class, %opts) = @_;

    # TODO: deprecated
    my @listeners;
    if (ref($opts{Listen}) eq "ARRAY") {
        @listeners = @{$opts{Listen}};
    } else {
        @listeners = ($opts{Listen} or ());
    }

    bless {
        auto_accept   => !$opts{ManualAccept},
        line_buffered => $opts{LineBuffered},
        buffered      => $opts{Buffered},
        auto_write    => !$opts{ManualWrite},
        auto_read     => !$opts{ManualRead},
        type          => $opts{Type} || "stream",
        listenfh      => { map { $_ => 1 } @listeners },
        readfh        => IO::Select->new(@listeners),
        writefh       => IO::Select->new(),
        fhs           => { },
        events        => [ ],
        actionq       => [ ],
    }, $class;
}

=head2 B<mux([$timeout])>

This method will block until ether an event occurs on one of the file handles
or the $timeout (floating point seconds) expires.  If the $timeout argument is
not present, it waits forever.  If $timeout is 0, it returns immediately.

The return value is always a hash, which always has the key 'type', indicating
what kind it is.  It will also usually carry the 'fh' key, indicating what file
handle the event happened on.

The 'type' key can have the following values:

=over

=item timeout

Nothing happened and timeout occurred.

=item error

The C<select()> system call failed. The event hash will have the key 'error',
which is set to the value of $! at the time of the error.

=item connect

A new client connected to a listening socket.

=item accept

A new client is trying to connect to a listening socket, but the user code must
call accept manually.  This only happens when the ManualAccept option is
set.

=item read

A socket has incoming data.  If the socket's LineBuffered option is set, this
will always be a full line, including the terminating newline.  The data is
contained in the 'data' key of the event hash.  If recv() returned a sender
address, it is contained in the 'sender' key and must be manually unpacked
according to the socket domain, e.g. with C<Socket::unpack_sockaddr_in()>.

=item disconnect

A socket disconnected, or a pipe was closed.

=item can_write

The ManualWrite option is set for the file handle, and C<select()> has
indicated that the handle can be written to.

=item can_read

The ManualRead option is set for the file handle, and C<select()> has
indicated that the handle can be read from.

=back

=cut

sub mux {
    my $self = shift;
    my $event;

    # We call get_event in a loop, because it may not always return an event
    # even when a socket is readable (e.g. if a half line is sent on a
    # LineBuffered socket). This means that it may take arbitrarily long to
    # return, regardless of the timeout parameter, because the same timeout is
    # reused every time.
    until ($event = $self->get_event(@_)) {}
    return $event;
}

sub get_event {
    my ($self, $timeout) = @_;

    # pending events?
    if (my $event = shift @{$self->{events}}) {
        return $event;
    }

    # actions to execute?
    while (my $action = shift @{$self->{actionq}}) {
        $action->($self);
    }

    # timeouts to respect?
    my $time = time;
    my $timeout_fh = undef;
    for my $fh ($self->{readfh}->handles) {
        if (my $abs_timeout = $self->{fhs}{$fh}{abs_timeout}) {
            my $rel_timeout = $abs_timeout - $time;

            if (!defined $timeout || $rel_timeout < $timeout) {
                $timeout = $rel_timeout;
                $timeout_fh = $fh;
            }
        }
    }

    $! = 0;
    # TODO : handle OOB data
    my @result = IO::Select->select($self->{readfh}, $self->{writefh},
        undef, (!defined $timeout || $timeout > 0 ? $timeout : 0));

    if (@result == 0) {
        if ($!) {
            return { type => 'error', error => $! };
        } elsif ($timeout_fh) {
            return { type => 'timeout', fh => $timeout_fh };
        } else {
            return { type => 'timeout' };
        }
    }

    # buffers to flush?, can_write is set.
    for my $fh (@{$result[1]}) {
        if(exists $self->{fhs}{$fh}{connected}
            and $self->{fhs}{$fh}{connected} == 0) {
            $self->{fhs}{$fh}{connected} = 1;

            my $packederror = getsockopt($fh, SOL_SOCKET, SO_ERROR);

            if(!defined $packederror) {
                $self->push_event({ type => 'connected', fh => $fh });
            } else {
                my $error = unpack("i", $packederror);
                if($error == 0) {
                    $self->push_event({ type => 'connected', fh => $fh });
                } else {
                    $self->push_event({ type => 'error',
                            fh => $fh, error=> $error });
                    $self->disconnect($fh);
                }
            }

        } elsif ($self->{fhs}{$fh}{auto_write}) {
            $self->send($fh);
        } else {
            $self->push_event({ type => 'can_write', fh => $fh });
        }
    }

    # incoming data?
    for my $fh (@{$result[0]}) {
        delete $self->{fhs}{$fh}{abs_timeout};

        if ($self->{listenfh}{$fh}) {
            # new connection
            if ($self->{auto_accept}) {
                my $client = $fh->accept or next;
                $self->push_event({ type => 'connect', fh => $client });
                $self->add($client);
            } else {
                $self->push_event({ type => 'accept', fh => $fh });
            }

        } elsif (!$self->{fhs}{$fh}{auto_read}) {
            $self->push_event({ type => 'can_read', fh => $fh });

        } else {
            $self->read_all($fh);
        }
    }

    return shift @{$self->{events}};
}

=head2 B<add($handle, [ %options ])>

Add a socket to the internal list of handles being watched.  If one of the
current listening sockets is given here, accept will be called on it.

To add a new listening socket, use C<add_listener()> instead.

The optional parameters for the handle will be taken from the IO::EventMux
object if not given here:

=over

=item LineBuffered

If true, only whole lines will be read from the file handle.

=item ManualWrite

By default, write buffering is handled automatically, which is fine for
low-throughput purposes.  If a lot of data needs to be sent over the file
handle, manual write buffering will give better performance and higher
flexibility.

In both cases you can use send() to write data to the file handle.

=item ManualRead

Never read or recv on the file handle. When the socket becomes readable, a
C<can_read> event is returned.

=item Type

Either "stream" or "dgram". Be sure to set this to "dgram" for UDP sockets.

=item Meta

An optional scalar piece of metadata for the file handle.  Can be retrieved and
manipulated later with meta()

=back

=cut

sub add {
    my ($self, $arg, %opts) = @_;

    my $client = ($self->{listenfh}{$arg} ? $arg->accept() : $arg)
        or return;

    $self->{readfh}->add($client);

    # If this is not UDP(SOCK_DGRAM) send a connect event.
    if (!UNIVERSAL::can($client, "socktype") or ($client->socktype||0) != SOCK_DGRAM) {
        $self->{writefh}->add($client);
        $self->{fhs}{$client}{connected} = 0;
    }

    $self->{fhs}{$client}{line_buffered} = (exists $opts{LineBuffered} ?
        $opts{LineBuffered} : $self->{line_buffered});
    
    $self->{fhs}{$client}{buffered} = (exists $opts{Buffered} ?
        $opts{Buffered} : $self->{buffered});

    $self->{fhs}{$client}{auto_write} = (exists $opts{ManualWrite} ?
        !$opts{ManualWrite} : $self->{auto_write});

    $self->{fhs}{$client}{auto_read} = (exists $opts{ManualRead} ?
        !$opts{ManualRead} : $self->{auto_read});

    $self->{fhs}{$client}{type} = (exists $opts{Type} ?
        $opts{Type} : $self->{type});

    if ($self->{fhs}{$client}{auto_read} || $self->{fhs}{$client}{auto_write}) {
        $self->nonblock($client);
    }

    if ($opts{Meta}) {
        $self->{fhs}{$client}{meta} = $opts{Meta};
    }
}

=head2 B<handles()>

Returns a list of file handles managed by this object.

=cut

sub handles {
    my ($self) = @_;
    return $self->{readfh}->handles;
}

=head2 B<meta($fh, [$newval])>

Set or get a piece of metadata on the filehandle.  This can be any scalar
value.

=cut

sub meta {
    my ($self, $fh, $newval) = @_;

    if (@_ > 2) {
        $self->{fhs}{$fh}{meta} = $newval;
    }
    return $self->{fhs}{$fh}{meta};
}

=head2 B<remove($fh)>

Make EventMux forget about a file handle. The caller will then take over the
responsibility of closing it.

=cut

sub remove {
    my ($self, $fh) = @_;

    $self->{readfh}->remove($fh);
    $self->{writefh}->remove($fh);
    delete $self->{listenfh}{$fh};
    delete $self->{fhs}{$fh};
}

=head2 B<disconnect($fh, [ $instant ])>

Close a file handle.  File handles managed by EventMux must be closed through
this method to make sure all resources are freed.

If $instant is not true, a 'disconnect' event will be returned by mux() before
actually closing the file handle.

=cut

sub disconnect {
    my ($self, $fh, $instant) = @_;

    return if $self->{fhs}{$fh}{disconnecting};

    $self->{readfh}->remove($fh);
    $self->{writefh}->remove($fh);
    $self->{fhs}{$fh}{disconnecting} = 1;
    delete $self->{listenfh}{$fh};

    if ($instant) {
        $self->close_fh($fh);
    } else {
        $self->push_event({ type => 'disconnect', fh => $fh });

        # wait with the close so a valid file handle can be returned
        push @{$self->{actionq}}, sub {
            $self->close_fh($fh);
        };
    }
}

=head2 B<close_fh($fh)>

Force the file handle to be closed. I<Must only> be called after a "disconnect"
event has been received from C<mux()>.

=cut

sub close_fh {
    my ($self, $fh) = @_;

    if ($self->{fhs}{$fh}) {
        delete $self->{fhs}{$fh};
        shutdown $fh, 2;
        close $fh or warn "closing $fh: $!";
    }
}

=head2 B<add_listener($socket)>

Register a new listening socket.

=cut

sub add_listener {
    my ($self, @socks) = @_;

    foreach my $sock (@socks) {
        $self->nonblock($sock);
        $self->{readfh}->add($sock);
        $self->{listenfh}{$sock} = 1;
    }
}

=head2 B<timeout($fh, [ $set_timeout ])>

Get or set the read timeout for this file handle in seconds. When no data has
been received for this many seconds, a 'timeout' event will be returned for this
file handle.

Set the undefined value for infinite timeout, which is the default.

The timeout is reset to infinity if any data is read from the file handle, so it
should be set again before the next read if desired.

=cut

sub timeout {
    my ($self, $fh, $new_val) = @_;

    my $time = time;

    if (@_ > 2) {
        if (!defined $new_val) {
            delete $self->{fhs}{$fh}{abs_timeout};
        } else {
            $self->{fhs}{$fh}{abs_timeout} = $new_val + $time;
        }
    }
    return $self->{fhs}{$fh}{abs_timeout} - $time;
}

=head2 B<buflen($fh)>

Queries the length of the output buffer for this file handle.  This only
applies if ManualWrite is turned off, which is the default. For Type="dgram"
sockets, it returns the number of datagrams in the queue.

An application can use this method to see whether it should send more data or
wait until the buffer queue is a bit shorter.

=cut

sub buflen {
    my ($self, $fh) = @_;
    my $meta = $self->{fhs}{$fh};

    if ($meta->{type} eq "dgram") {
        return $meta->{outbuffer} ? scalar(@{$meta->{outbuffer}}) : 0;
    }

    return $meta->{outbuffer} ? length($meta->{outbuffer}) : 0;
}

=head2 B<send($fh, @data)>

Attempt to write each item of @data to $fh. Can only be used when ManualWrite is
off (default).

=over

=item If the socket is of Type="stream" (default)

Returns true on success, undef on error.  The data may or may not have been
written to the socket when the function returns.  Therefore the socket should
not be closed until L</B<buflen($fh)>> returns 0.  If an unrecoverable error
occurs, the file handled will be closed.

=item If the socket is of Type="dgram"

Each item in @data will be sent as a separate packet.  Returns the number of
packets written to the socket, which may include packets that were already in
the queue. Returns undef on error, in which case the file handle is closed.

=back

=cut

sub send {
    my ($self, $fh, @data) = @_;
    return $self->sendto($fh, undef, @data);
}

=head2 B<sendto($fh, $to, @data)>

Like C<send()>, but with the recepient C<$to> as a packed sockaddr structure,
such as the one returned by C<Socket::pack_sockaddr_in()>. Only for Type="dgram"
sockets.

=cut

sub sendto {
    my ($self, $fh, $to, @data) = @_;

    if (not defined $fh) {
        carp "send() on an undefined file handle";
        return;
    }
    my $meta = $self->{fhs}{$fh} or return undef;
    return undef if $meta->{disconnecting};

    if (not $meta->{auto_write}) {
        carp "send() on a ManualWrite file handle";
        return;
    }

    if ($meta->{type} eq "dgram") {
        return $self->send_dgram($fh, $to, @data);
    } else {
        return $self->send_stream($fh, @data);
    }
}

sub send_dgram {
    my ($self, $fh, $all_to, @newdata) = @_;
    my $meta = $self->{fhs}{$fh} or return undef;

    push @{$meta->{outbuffer}}, map { [$_, $all_to] } @newdata;

    my $packets_sent = 0;

    while (my $queue_item = shift @{$meta->{outbuffer}}) {
        my ($data, $to) = @$queue_item;
        my $rv = $self->my_send($fh, $data, (defined $to ? $to : ()));

        if (!defined $rv) {
            if ($! == POSIX::EWOULDBLOCK) {
                # retry later
                unshift @{$meta->{outbuffer}}, $queue_item;
                $self->{writefh}->add($fh);
                return $packets_sent;
            } else {
                # error on socket
                warn "send: $!";
                $self->disconnect($fh);
            }
            return undef;

        } elsif ($rv < length $data) {
            cluck "Incomplete datagram sent (should not happen)";

        } else {
            # all pending data was sent
            $packets_sent++;
        }
    }

    $self->{writefh}->remove($fh);

    return $packets_sent;
}

sub send_stream {
    my ($self, $fh, @data) = @_;
    my $meta = $self->{fhs}{$fh} or return undef;

    # send pending data before this
    my $data = join "",
        (defined $meta->{outbuffer} ? $meta->{outbuffer} : ()), @data;

    delete $meta->{outbuffer};

    if (length $data == 0) {
        # no data to send
        $self->{writefh}->remove($fh);
        return 0;
    }

    my $rv = $self->my_send($fh, $data);
    
    if (!defined $rv) {
        if ($! == POSIX::EWOULDBLOCK or $! == POSIX::EAGAIN or !$self->{fhs}{$fh}{connected}) {
            # try later
            $meta->{outbuffer} = $data;
            $self->{writefh}->add($fh);
            return 0;
        } else {
            # error on socket
            warn "send: $!";
            $self->disconnect($fh);
        }
        return undef;

    } elsif ($rv < 0) {
        warn;
        $self->disconnect($fh);
        return undef;

    } elsif ($rv < length $data) {
        # only part of the data was sent
        $meta->{outbuffer} = substr $data, $rv;
        $self->{writefh}->add($fh);
    } else {
        # all pending data was sent
        $self->{writefh}->remove($fh);
    }

    return $rv;
}

sub my_send {
    my ($self, $fh, $data, @to) = @_;

    my $rv;
    $! = undef;
    if (UNIVERSAL::can($fh, "send") and !$fh->isa("IO::Socket::SSL")) {
        $rv = eval { $fh->send($data, 0, @to) };
    } else {
        $rv = eval { syswrite $fh, $data };
    }
    return $@ ? undef : $rv;
}

sub select_write {
    $_[0]->{writefh}->add($_[1]);
}

sub unselect_write {
    $_[0]->{writefh}->remove($_[1]);
}

sub push_event {
    push @{$_[0]->{events}}, $_[1];
}

# nonblock($socket) puts socket into nonblocking mode
sub nonblock {
    my $socket = $_[1];

    my $flags = fcntl($socket, F_GETFL, 0)
        or die "Can't get flags for socket: $!\n";
    if (not $flags & O_NONBLOCK) {
        fcntl($socket, F_SETFL, $flags | O_NONBLOCK)
            or die "Can't make socket nonblocking: $!\n";
    }
}

# Keeps reading from a file handle until POSIX::EWOULDBLOCK is returned
sub read_all {
    my ($self, $fh) = @_;
    my $cfg = $self->{fhs}{$fh};

    while (1) {
        my ($data, $rv, $sender) = ('');
        if (UNIVERSAL::can($fh, "recv") and !$fh->isa("IO::Socket::SSL")) {
            $rv = $fh->recv($data, 65536, 0);
            $sender = $rv if defined $rv && $rv ne "";
        } else {
            $rv = sysread $fh, $data, 65536;
        }

        if (not defined $rv) {
            if ($! != POSIX::EWOULDBLOCK) {
                $self->push_event({ type => 'error', error => $!, fh => $fh });
            }
            return;
        }

        if (length $data == 0 and $cfg->{type} eq "stream") {
            # client disconnected
            $self->disconnect($fh);
            return;
        }

        my %event = (type => 'read', fh => $fh);
        $event{'sender'} = $sender if defined $sender;

        if($cfg->{buffered}) {
            my ($type, @args) = @{$cfg->{buffered}};
            $cfg->{inbuffer} .= $data;
            
            if($type eq "Size"){
                my ($pattern, $offset) = (@args, 0);
                my $length = 
                    (unpack($pattern, $cfg->{inbuffer}))[0];
                my $datastart = length(pack($pattern, $length));

                while($length <= length($cfg->{inbuffer})) {
                    my %copy = %event;
                    $copy{'data'} = substr($cfg->{inbuffer},
                        $datastart, $length+$offset);
                    substr($cfg->{inbuffer}, $datastart, $length+$offset) = '';
                    $self->push_event(\%copy);
                }
            }
            
        } elsif ($cfg->{line_buffered}) {
            $cfg->{inbuffer} .= $data;
            while ($cfg->{inbuffer} =~ s/(.*\n)//) {
                my %copy = %event;
                $copy{'data'} = $1;
                $self->push_event(\%copy);
            }

        } else {
            $event{'data'} = $data;
            $self->push_event(\%event);
        }
    }
}

1;

=head1 AUTHOR

Jonas Jensen <jbj@knef.dk>
Troels Liebe Bentsen <tlb@rapanden.dk>

=cut

# vim: et sw=4 sts=4 tw=80
